File: ALODWebUtility/TabNavigator/TabItemList.vb
Namespace TabNavigation

    <Serializable()>
    Public Class TabItemList
        Implements IList(Of TabItem)

        Protected _list As New List(Of TabItem)

        Public ReadOnly Property Count() As Integer Implements System.Collections.Generic.ICollection(Of TabItem).Count
            Get
                Return _list.Count
            End Get
        End Property

        Public ReadOnly Property IsReadOnly() As Boolean Implements System.Collections.Generic.ICollection(Of TabItem).IsReadOnly
            Get
                Return False
            End Get
        End Property

        Default Public Property Item(ByVal index As Integer) As TabItem Implements System.Collections.Generic.IList(Of TabItem).Item
            Get
                Return _list(index)
            End Get
            Set(ByVal value As TabItem)
                _list(index) = value
            End Set
        End Property

        Default Public Property Item(ByVal page As String) As TabItem
            Get
                For Each row As TabItem In _list
                    If (String.Equals(row.Page, page, StringComparison.InvariantCultureIgnoreCase)) Then
                        Return row
                    End If
                Next
                Return Nothing
            End Get
            Set(ByVal value As TabItem)
                For Each row As TabItem In _list
                    If (row.Page.CompareTo(page)) Then
                        row = value
                    End If
                Next
            End Set
        End Property

        Public Sub Add(ByVal item As TabItem) Implements System.Collections.Generic.ICollection(Of TabItem).Add
            _list.Add(item)
        End Sub

        Public Sub Clear() Implements System.Collections.Generic.ICollection(Of TabItem).Clear
            _list.Clear()
        End Sub

        Public Function Contains(ByVal item As TabItem) As Boolean Implements System.Collections.Generic.ICollection(Of TabItem).Contains
            Return _list.Contains(item)
        End Function

        Public Sub CopyTo(ByVal array() As TabItem, ByVal arrayIndex As Integer) Implements System.Collections.Generic.ICollection(Of TabItem).CopyTo
            _list.CopyTo(array, arrayIndex)
        End Sub

        Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of TabItem) Implements System.Collections.Generic.IEnumerable(Of TabItem).GetEnumerator
            Return _list.GetEnumerator()
        End Function

        Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            Return _list.GetEnumerator()
        End Function

        Public Function IndexOf(ByVal item As TabItem) As Integer Implements System.Collections.Generic.IList(Of TabItem).IndexOf
            Return _list.IndexOf(item)
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal item As TabItem) Implements System.Collections.Generic.IList(Of TabItem).Insert
            _list.Insert(index, item)
        End Sub

        Public Function Remove(ByVal item As TabItem) As Boolean Implements System.Collections.Generic.ICollection(Of TabItem).Remove
            Return _list.Remove(item)
        End Function

        Public Sub RemoveAt(ByVal index As Integer) Implements System.Collections.Generic.IList(Of TabItem).RemoveAt
            _list.RemoveAt(index)
        End Sub

    End Class

End Namespace

File: ALODWebUtility.CS/TabNavigator/TabItemList.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace ALODWebUtility.TabNavigation
{
    [Serializable]
    public class TabItemList : IList<TabItem>
    {
        protected List<TabItem> _list = new List<TabItem>();

        public int Count
        {
            get
            {
                return _list.Count;
            }
        }

        public bool IsReadOnly
        {
            get
            {
                return false;
            }
        }

        public TabItem this[int index]
        {
            get
            {
                return _list[index];
            }
            set
            {
                _list[index] = value;
            }
        }

        public TabItem this[string page]
        {
            get
            {
                foreach (TabItem row in _list)
                {
                    if (string.Equals(row.Page, page, StringComparison.InvariantCultureIgnoreCase))
                    {
                        return row;
                    }
                }
                return null;
            }
            set
            {
                for (int i = 0; i < _list.Count; i++)
                {
                    if (_list[i].Page.CompareTo(page) == 0)
                    {
                        _list[i] = value;
                    }
                }
            }
        }

        public void Add(TabItem item)
        {
            _list.Add(item);
        }

        public void Clear()
        {
            _list.Clear();
        }

        public bool Contains(TabItem item)
        {
            return _list.Contains(item);
        }

        public void CopyTo(TabItem[] array, int arrayIndex)
        {
            _list.CopyTo(array, arrayIndex);
        }

        public IEnumerator<TabItem> GetEnumerator()
        {
            return _list.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _list.GetEnumerator();
        }

        public int IndexOf(TabItem item)
        {
            return _list.IndexOf(item);
        }

        public void Insert(int index, TabItem item)
        {
            _list.Insert(index, item);
        }

        public bool Remove(TabItem item)
        {
            return _list.Remove(item);
        }

        public void RemoveAt(int index)
        {
            _list.RemoveAt(index);
        }
    }
}
